File: frontend/packages/evaluationMetadata.js
export const EVALUATION_METADATA_SCHEMA_VERSION = 1;
export const MAX_EVALUATION_METADATA_BYTES = 64 * 1024;

function decodeMetadata(value) {
  let metadata;
  try {
    metadata = typeof value === "string"
      ? (value.trim() ? JSON.parse(value) : {})
      : value ?? {};
  } catch {
    throw new Error("Metadata đánh giá không phải JSON hợp lệ");
  }
  if (!metadata || typeof metadata !== "object" || Array.isArray(metadata)) {
    throw new Error("Metadata đánh giá phải là object");
  }
  return structuredClone(metadata);
}

function assertMetadataSize(metadata) {
  const encoded = new TextEncoder().encode(JSON.stringify(metadata));
  if (encoded.byteLength > MAX_EVALUATION_METADATA_BYTES) {
    throw new Error("Metadata đánh giá vượt giới hạn 64 KiB");
  }
}

export function migrateEvaluationMetadata(value) {
  const metadata = decodeMetadata(value);
  const version = metadata.schemaVersion;
  if (version === undefined || version === null || version === 0) {
    metadata.schemaVersion = EVALUATION_METADATA_SCHEMA_VERSION;
  } else if (version !== EVALUATION_METADATA_SCHEMA_VERSION) {
    throw new Error(`Không hỗ trợ schemaVersion metadata đánh giá: ${version}`);
  }
  assertMetadataSize(metadata);
  return metadata;
}

export function parseEvaluationMetadataStrict(value) {
  return migrateEvaluationMetadata(value);
}

export function parseEvaluationMetadataForDisplay(value) {
  try {
    return {
      metadata: parseEvaluationMetadataStrict(value),
      error: null,
      canPersist: true,
      raw: value,
    };
  } catch (error) {
    return {
      metadata: { schemaVersion: EVALUATION_METADATA_SCHEMA_VERSION },
      error,
      canPersist: false,
      raw: value,
    };
  }
}

export function parseEvaluationMetadata(value) {
  return parseEvaluationMetadataStrict(value);
}

export function serializeEvaluationMetadata(value) {
  return JSON.stringify(parseEvaluationMetadataStrict(value));
}

export const parseStrict = parseEvaluationMetadataStrict;
export const parseForDisplay = parseEvaluationMetadataForDisplay;
export const serialize = serializeEvaluationMetadata;
export const migrate = migrateEvaluationMetadata;

File: backend/sync/evaluation_metadata.py
import json


EVALUATION_METADATA_SCHEMA_VERSION = 1
MAX_EVALUATION_METADATA_BYTES = 64 * 1024


def _decode_evaluation_metadata(value):
    if value in (None, ""):
        return {}
    try:
        metadata = json.loads(value) if isinstance(value, str) else value
    except (TypeError, ValueError, json.JSONDecodeError) as exc:
        raise ValueError("Evaluation metadata must be valid JSON.") from exc
    if not isinstance(metadata, dict):
        raise ValueError("Evaluation metadata must be a JSON object.")
    # Metadata is a JSON contract. Round-tripping creates an isolated plain
    # object and rejects values that could never be persisted as JSON.
    try:
        return json.loads(json.dumps(metadata, ensure_ascii=False))
    except (TypeError, ValueError) as exc:
        raise ValueError("Evaluation metadata must contain JSON values only.") from exc


def _validate_metadata_size(metadata):
    encoded = json.dumps(
        metadata, ensure_ascii=False, separators=(",", ":")
    ).encode("utf-8")
    if len(encoded) > MAX_EVALUATION_METADATA_BYTES:
        raise ValueError("Evaluation metadata exceeds the 64 KiB limit.")


def migrate_evaluation_metadata(value):
    metadata = _decode_evaluation_metadata(value)
    version = metadata.get("schemaVersion")
    if version in (None, 0):
        metadata["schemaVersion"] = EVALUATION_METADATA_SCHEMA_VERSION
    elif version != EVALUATION_METADATA_SCHEMA_VERSION:
        raise ValueError("Evaluation metadata schemaVersion is not supported.")
    _validate_metadata_size(metadata)
    return metadata


def parse_evaluation_metadata(value, *, require_version=True):
    # ``require_version`` remains for caller compatibility. Missing/version-0
    # legacy metadata is migrated; unknown future schemas always fail closed.
    del require_version
    return migrate_evaluation_metadata(value)


def parse_evaluation_metadata_for_display(value):
    try:
        return {
            "metadata": parse_evaluation_metadata(value),
            "error": None,
            "canPersist": True,
            "raw": value,
        }
    except ValueError as exc:
        return {
            "metadata": {
                "schemaVersion": EVALUATION_METADATA_SCHEMA_VERSION,
            },
            "error": exc,
            "canPersist": False,
            "raw": value,
        }


def dump_evaluation_metadata(metadata):
    normalized = parse_evaluation_metadata(metadata)
    return json.dumps(normalized, ensure_ascii=False, separators=(",", ":"))
